xclim/core/units.py

```diff
@@ -285,60 +287,79 @@ def convert_units_to(
 
     Returns
     -------
-    xr.DataArray or float or int or str or Any
+    str or xr.DataArray or units.Unit or units.Quantity
         The source value converted to target's units.
+        The outputted type is always similar to `source` initial type.
     """
     # Target units
     if isinstance(target, units.Unit):
-        tu = target
+        target_unit = target
     elif isinstance(target, (str, xr.DataArray)):
-        tu = units2pint(target)
+        target_unit = units2pint(target)
     else:
-        raise NotImplementedError
+        raise NotImplementedError(
+            "target must be either a pint Unit or a xarray DataArray."
+        )
 
     if isinstance(source, str):
         q = str2pint(source)
         # Return magnitude of converted quantity. This is going to fail if units are not compatible.
-        return q.to(tu).m
+        return q.to(target_unit).m
 
     if isinstance(source, units.Quantity):
-        return source.to(tu).m
+        return source.to(target_unit).m
 
     if isinstance(source, xr.DataArray):
-        fu = units2pint(source)
-        tu_u = pint2cfunits(tu)
+        source_unit = units2pint(source)
+        target_cf_unit = pint2cfunits(target_unit)
 
-        if fu == tu:
+        if source_unit == target_unit:
             # The units are the same, but the symbol may not be.
-            source.attrs["units"] = tu_u
+            source.attrs["units"] = target_cf_unit
             return source
 
         with units.context(context or "none"):
+            if (
+                _is_precipitation_amount(source, source_unit)
+                and target_unit.dimensionality.get("[time]") == -1
+            ):
+                source = amount2rate(source)
```

Comment:
I realized a potential bug in `amount2rate` yesterday.
If the time axis is not continuous, the sampling rate is determined by the time diff. This means that if you have a large gap in your data, the data point at the start of that gap gets divided by a very large time period. This is silent.
Thus, doing this conversion automatically seems dangerous to me! The solution could be that `amount2rate` raises if `xr.infer_freq` fails on the data, unless the behaviour is deactivate by some argument to be added.
What do you think?

Reply:
I agree with this. I was surprised to realize the function was so liberal.

Reply:
Fixed in [d1fea25](https://github.com/Ouranosinc/xclim/pull/1206/commits/d1fea25b72dc1dbf715e5b39db0ff88971cdb937)